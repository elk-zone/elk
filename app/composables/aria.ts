export type AriaLive = 'off' | 'polite' | 'assertive'
export type AriaAnnounceType = 'announce' | 'mute' | 'unmute'

const ariaAnnouncer = useEventBus<AriaAnnounceType, string | undefined>(Symbol('aria-announcer'))

export function useAriaAnnouncer() {
  const announce = (message: string) => {
    ariaAnnouncer.emit('announce', message)
  }

  const mute = () => {
    ariaAnnouncer.emit('mute')
  }

  const unmute = () => {
    ariaAnnouncer.emit('unmute')
  }

  return { announce, ariaAnnouncer, mute, unmute }
}

export function useAriaLog() {
  const logs = ref<any[]>([])

  const announceLogs = (messages: any[]) => {
    logs.value = messages
  }

  const appendLogs = (messages: any[]) => {
    logs.value = logs.value.concat(messages)
  }

  const clearLogs = () => {
    logs.value = []
  }

  return {
    announceLogs,
    appendLogs,
    clearLogs,
    logs,
  }
}

export function useAriaStatus() {
  const status = ref<any>('')

  const announceStatus = (message: any) => {
    status.value = message
  }

  const clearStatus = () => {
    status.value = ''
  }

  return {
    announceStatus,
    clearStatus,
    status,
  }
}
