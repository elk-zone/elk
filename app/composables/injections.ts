import { InjectionKeyDropdownContext } from '~/constants/symbols'

export function useDropdownContext() {
  return inject(InjectionKeyDropdownContext, undefined)
}
