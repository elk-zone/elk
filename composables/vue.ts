import type { ComponentInternalInstance } from 'vue'
import { onActivated, onDeactivated, ref } from 'vue'
import type { ActiveHeadEntry, HeadEntryOptions, UseHeadInput } from '@vueuse/head'
import type { HeadAugmentations } from '@nuxt/schema'
import { useHead } from '#head'

export const isHydrated = ref(false)

export const onHydrated = (cb: () => unknown) => {
  watchOnce(isHydrated, () => cb(), { immediate: isHydrated.value })
}

/**
 * ### Whether the current component is running in the background
 *
 * for handling problems caused by the keepalive function
 */
export function useDeactivated() {
  const deactivated = ref(false)
  onActivated(() => deactivated.value = false)
  onDeactivated(() => deactivated.value = true)

  return deactivated
}

/**
 * ### When the component is restored from the background
 *
 * for handling problems caused by the keepalive function
 */
export function onReactivated(hook: Function, target?: ComponentInternalInstance | null): void {
  const initial = ref(true)
  onActivated(() => {
    if (initial.value)
      return
    hook()
  }, target)
  onDeactivated(() => initial.value = false)
}

// TODO: Workaround for Nuxt bug: https://github.com/elk-zone/elk/pull/199#issuecomment-1329771961
export function useHeadFixed<T extends HeadAugmentations>(input: UseHeadInput<T>, options?: HeadEntryOptions): ActiveHeadEntry<UseHeadInput<T>> | void {
  const deactivated = useDeactivated()
  if (input && typeof input === 'object' && !('value' in input)) {
    const title = 'title' in input ? input.title : undefined
    if (process.server && title) {
      input.meta = input.meta || []
      if (Array.isArray(input.meta)) {
        input.meta.push(
          { property: 'og:title', content: (typeof input.title === 'function' ? input.title() : input.title) as string },
        )
      }
    }
    else if (title) {
      (input as any).title = () => isHydrated.value ? typeof title === 'function' ? title() : title : ''
    }
  }
  return useHead(() => {
    if (deactivated.value)
      return {}
    return resolveUnref(input)
  }, options)
}
