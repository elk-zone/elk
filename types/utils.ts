export type Mutable<T> = {
  -readonly[P in keyof T]: T[P]
}

export type Overwrite<T, O> = Omit<T, keyof O> & O
export type MarkNonNullable<T, K extends keyof T> = Overwrite<T, {
  [P in K]-?: NonNullable<T[P]>
}>
