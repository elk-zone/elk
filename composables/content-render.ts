import { TEXT_NODE } from 'ultrahtml'
import type { Node } from 'ultrahtml'
import { Fragment, h, isVNode } from 'vue'
import type { VNode } from 'vue'
import { RouterLink } from 'vue-router'
import { decode } from 'tiny-decode'
import type { ContentParseOptions } from './content-parse'
import { parseMastodonHTML } from './content-parse'
import ContentCode from '~/components/content/ContentCode.vue'
import ContentMentionGroup from '~/components/content/ContentMentionGroup.vue'
import AccountHoverWrapper from '~/components/account/AccountHoverWrapper.vue'

function getTexualAstComponents(astChildren: Node[]): string {
  return astChildren
    .filter(({ type }) => type === TEXT_NODE)
    .map(({ value }) => value)
    .reduce((accumulator, current) => accumulator + current, '')
    .trim()
}

/**
* Raw HTML to VNodes
*/
export function contentToVNode(
  content: string,
  options?: ContentParseOptions,
): VNode {
  let tree = parseMastodonHTML(content, options)

  const textContents = getTexualAstComponents(tree.children)

  // if the username only contains emojis, we should probably show the emojis anyway to avoid a blank name
  if (options?.hideEmojis && textContents.length === 0)
    tree = parseMastodonHTML(content, { ...options, hideEmojis: false })

  return h(Fragment, (tree.children as Node[] || []).map(n => treeToVNode(n)))
}

function nodeToVNode(node: Node): VNode | string | null {
  if (node.type === TEXT_NODE)
    return node.value

  if (node.name === 'mention-group')
    return h(ContentMentionGroup, node.attributes, () => node.children.map(treeToVNode))

  if ('children' in node) {
    if (node.name === 'a' && (node.attributes.href?.startsWith('/') || node.attributes.href?.startsWith('.'))) {
      node.attributes.to = node.attributes.href
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { href, target, ...attrs } = node.attributes
      return h(
        RouterLink as any,
        attrs,
        () => node.children.map(treeToVNode),
      )
    }
    return h(
      node.name,
      node.attributes,
      node.children.map(treeToVNode),
    )
  }
  return null
}

function treeToVNode(
  input: Node,
): VNode | string | null {
  if (input.type === TEXT_NODE)
    return decode(input.value)

  if ('children' in input) {
    const node = handleNode(input)
    if (node == null)
      return null
    if (isVNode(node))
      return node
    return nodeToVNode(node)
  }
  return null
}

function handleMention(el: Node) {
  // Redirect mentions to the user page
  if (el.name === 'a' && el.attributes.class?.includes('mention')) {
    const href = el.attributes.href
    if (href) {
      const matchUser = href.match(UserLinkRE)
      if (matchUser) {
        const [, server, username] = matchUser
        const handle = `${username}@${server.replace(/(.+\.)(.+\..+)/, '$2')}`
        el.attributes.href = `/${server}/@${username}`
        return h(AccountHoverWrapper, { handle, class: 'inline-block' }, () => nodeToVNode(el))
      }
      const matchTag = href.match(TagLinkRE)
      if (matchTag) {
        const [, , name] = matchTag
        el.attributes.href = `/${currentServer.value}/tags/${name}`
      }
    }
  }
  return undefined
}

function handleCodeBlock(el: Node) {
  if (el.name === 'pre' && el.children[0]?.name === 'code') {
    const codeEl = el.children[0] as Node
    const classes = codeEl.attributes.class as string
    const lang = classes?.split(/\s/g).find(i => i.startsWith('language-'))?.replace('language-', '')
    const code = codeEl.children[0] ? treeToText(codeEl.children[0]) : ''
    return h(ContentCode, { lang, code: encodeURIComponent(code) })
  }
}

function handleNode(el: Node) {
  return handleCodeBlock(el) || handleMention(el) || el
}
