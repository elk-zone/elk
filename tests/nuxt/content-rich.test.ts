import { describe, expect, it, vi } from 'vitest'
import { renderToString } from 'vue/server-renderer'
import { format } from 'prettier'
import type { mastodon } from 'masto'
import { mockComponent } from 'nuxt-vitest/utils'
import { contentToVNode } from '~/composables/content-render'
import type { ContentParseOptions } from '~/composables/content-parse'

describe('content-rich', () => {
  it('empty', async () => {
    const { formatted } = await render('')
    expect(formatted).toMatchSnapshot()
  })

  it('plain text', async () => {
    const { formatted } = await render('hello there', { collapseMentionLink: true })
    expect(formatted).toMatchSnapshot()
  })

  it('link + mention', async () => {
    // https://fosstodon.org/@ayo/109383002937620723
    const { formatted } = await render('<p>Happy ðŸ¤— weâ€™re now using <span class="h-card"><a href="https://webtoo.ls/@vitest" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>vitest</span></a></span> (migrated from chai+mocha) <a href="https://github.com/ayoayco/astro-reactive-library/pull/203" rel="nofollow noopener noreferrer" target="_blank"><span class="invisible">https://</span><span class="ellipsis">github.com/ayoayco/astro-react</span><span class="invisible">ive-library/pull/203</span></a></p>')
    expect(formatted).toMatchSnapshot()
  })

  it ('block with backticks', async () => {
    const { formatted } = await render('<p>```<br />[(`number string) (`tag string)]<br />```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('group mention', async () => {
    const { formatted } = await render('<p><span class="h-card"><a href="https://lemmy.ml/c/pilipinas" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>pilipinas</span></a></span></p>', {
      mentions: [{ id: '', username: 'pilipinas', url: 'https://lemmy.ml/c/pilipinas', acct: 'pilipinas@lemmy.ml' }],
    })
    expect(formatted).toMatchSnapshot('html')
  })

  it('inline code with link', async () => {
    const { formatted } = await render('<p>Inline code with link: `<a href="https://api.iconify.design/noto.css?icons=1st-place-medal,2nd-place-medal" target="_blank" rel="nofollow noopener noreferrer" class="status-link unhandled-link" title="https://api.iconify.design/noto.css?icons=1st-place-medal,2nd-place-medal"><span class="invisible">https://</span><span class="ellipsis">api.iconify.design/noto.css?ic</span><span class="invisible">ons=1st-place-medal,2nd-place-medal</span></a>`</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('handles html within code blocks', async () => {
    const { formatted } = await render('<p>HTML block code:<br>```html<br>&lt;span class="icon--noto icon--noto--1st-place-medal"&gt;&lt;/span&gt;<br>&lt;span class="icon--noto icon--noto--2nd-place-medal-medal"&gt;&lt;/span&gt;<br>```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('handles formatting from servers', async () => {
    const { formatted } = await render('<h1>Fedi HTML Support Survey</h1><p>Does the following formatting come through accurately for you?</p><ul><li>This is an indented bulleted list (not just asterisks).</li><li><strong>This line is bold.</strong></li><li><em>This line is italic.</em></li></ul><ol><li>This list...</li><li>...is numbered and indented</li></ol><h1>This line is larger.</h1>')
    expect(formatted).toMatchSnapshot()
  })

  it('custom emoji', async () => {
    const { formatted } = await render('Daniel Roe :nuxt:', {
      emojis: {
        nuxt: {
          shortcode: 'nuxt',
          url: 'https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png',
          staticUrl: 'https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png',
          visibleInPicker: true,
        },
      },
    })
    expect(formatted).toMatchSnapshot()
  })

  it('code frame', async () => {
    // https://webtoo.ls/@antfu/109396489827394721
    const { formatted } = await render('<p>Testing code block</p><p>```ts<br />import { useMouse, usePreferredDark } from &#39;@vueuse/core&#39;</p><p>// tracks mouse position<br />const { x, y } = useMouse()</p><p>// is the user prefers dark theme<br />const isDark = usePreferredDark()<br />```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame 2', async () => {
    const { formatted } = await render('<p><span class=\"h-card\"><a href=\"https://webtoo.ls/@antfu\" class=\"u-url mention\">@<span>antfu</span></a></span> Testing<br />```ts<br />const a = hello<br />```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame no lang', async () => {
    const { formatted } = await render('<p>```<br />hello world<br />```<br />no lang</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame empty', async () => {
    const { formatted } = await render('<p>```<br /><br />```<br /></p>')
    expect(formatted).toMatchSnapshot()
  })

  it('collapse mentions', async () => {
    const { formatted } = await render('<p><span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>elk</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>elk</span></a></span> content <span class="h-card"><a href="https://m.webtoo.ls/@antfu" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>antfu</span></a></span> <span class="h-card"><a href="https://mastodon.roe.dev/@daniel" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>daniel</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@sxzz" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>sxzz</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@patak" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>patak</span></a></span> content</p>', {
      collapseMentionLink: true,
    })
    expect(formatted).toMatchInlineSnapshot(`
      "<p>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@elk\\"
          ></a
        ></span>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@elk\\"
          ></a
        ></span>
        content
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@antfu\\"
          ></a
        ></span>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/mastodon.roe.dev/@daniel\\"
          ></a
        ></span>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@sxzz\\"
          ></a
        ></span>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@patak\\"
          ></a
        ></span>
        content
      </p>
      "
    `)
  })

  it('hides collapsed mentions', async () => {
    const { formatted } = await render('<p><span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>elk</span></a></span> content</p>', {
      collapseMentionLink: true,
      inReplyToStatus: { account: { acct: 'elk@webtoo.ls' }, mentions: [] as mastodon.v1.StatusMention[] } as mastodon.v1.Status,
    })
    expect(formatted).toMatchInlineSnapshot(`
      "<p>content</p>
      "
    `)
  })

  it('shows some collapsed mentions inline', async () => {
    const { formatted } = await render('<p><span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>elk</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@antfu" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>antfu</span></a></span> content</p>', {
      collapseMentionLink: true,
      inReplyToStatus: { account: { acct: 'elk@webtoo.ls' }, mentions: [] as mastodon.v1.StatusMention[] } as mastodon.v1.Status,
    })
    expect(formatted).toMatchInlineSnapshot(`
      "<p>
        <span class=\\"h-card\\"
          ><a
            class=\\"u-url mention\\"
            rel=\\"nofollow noopener noreferrer\\"
            to=\\"/m.webtoo.ls/@antfu\\"
          ></a
        ></span>
        content
      </p>
      "
    `)
  })

  it('shows some collapsed mentions grouped', async () => {
    const { formatted } = await render('<p><span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>elk</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@antfu" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>antfu</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@patak" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>patak</span></a></span> <span class="h-card"><a href="https://m.webtoo.ls/@sxzz" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>sxzz</span></a></span>content</p>', {
      collapseMentionLink: true,
      inReplyToStatus: { account: { acct: 'elk@webtoo.ls' }, mentions: [] as mastodon.v1.StatusMention[] } as mastodon.v1.Status,
    })
    expect(formatted).toMatchInlineSnapshot(`
      "<p>
        <mention-group
          ><span class=\\"h-card\\"
            ><a
              class=\\"u-url mention\\"
              rel=\\"nofollow noopener noreferrer\\"
              to=\\"/m.webtoo.ls/@antfu\\"
            ></a
          ></span>
          <span class=\\"h-card\\"
            ><a
              class=\\"u-url mention\\"
              rel=\\"nofollow noopener noreferrer\\"
              to=\\"/m.webtoo.ls/@patak\\"
            ></a
          ></span>
          <span class=\\"h-card\\"
            ><a
              class=\\"u-url mention\\"
              rel=\\"nofollow noopener noreferrer\\"
              to=\\"/m.webtoo.ls/@sxzz\\"
            ></a></span></mention-group
        >content
      </p>
      "
    `)
  })

  it ('block with injected html, without language', async () => {
    const { formatted } = await render(`
      <pre>
        <code>
          &lt;a href="javascript:alert(1)">click me&lt;/a>
        </code>
      </pre>
    `)
    expect(formatted).toMatchSnapshot()
  })

  it ('block with injected html, with an unknown language', async () => {
    const { formatted } = await render(`
      <pre>
        <code class="language-xyzzy">
          &lt;a href="javascript:alert(1)">click me&lt;/a>
        </code>
      </pre>
    `)
    expect(formatted).toMatchSnapshot()
  })

  it ('block with injected html, with a known language', async () => {
    const { formatted } = await render(`
      <pre>
        <code class="language-js">
          &lt;a href="javascript:alert(1)">click me&lt;/a>
        </code>
      </pre>
    `)
    expect(formatted).toMatchSnapshot()
  })
})

describe('editor', () => {
  it('transform mentions', () => {
    const ast = parseMastodonHTML('<p><span class="h-card"><a href="https://m.webtoo.ls/@elk" class="u-url mention">@<span>elk</span></a></span> Hello</p>')
    const transformed = treeToText(ast)
    expect(transformed).toMatchSnapshot()
  })
})

async function render(content: string, options?: ContentParseOptions) {
  const vnode = contentToVNode(content, options)
  const html = (await renderToString(vnode))
    .replace(/<!--[\[\]]-->/g, '')
  let formatted = ''

  try {
    formatted = format(html, {
      parser: 'html',
    })
  }
  catch (e) {
    formatted = html
  }

  return {
    vnode,
    html,
    formatted,
  }
}

// mocks
vi.mock('vue-router', async () => {
  const { defineComponent, h } = await import('vue')
  return {
    RouterLink: defineComponent((attrs) => {
      return () => h('a', attrs)
    }),
  }
})

vi.mock('shiki-es', async (importOriginal) => {
  const mod = await importOriginal()
  return {
    ...(mod as any),
    setCDN() {},
  }
})

mockComponent('ContentMentionGroup', {
  setup(props, { slots }) {
    return () => h('mention-group', null, { default: () => slots?.default?.() })
  },
})

mockComponent('AccountHoverWrapper', {
  props: ['handle', 'class'],
  setup(_, { slots }) {
    return () => slots?.default?.()
  },
})
