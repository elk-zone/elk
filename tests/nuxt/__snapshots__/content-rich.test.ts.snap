// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`content-rich > block with backticks 1`] = `"<p><pre class="code-block">[(\`number string) (\`tag string)]</pre></p>"`;

exports[`content-rich > block with injected html, with a known language 1`] = `
"<pre>
        <code class="language-js">
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > block with injected html, with an unknown language 1`] = `
"<pre>
        <code class="language-xyzzy">
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > block with injected html, without language 1`] = `
"<pre>
        <code>
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > code frame 1`] = `
"<p>Testing code block</p><p></p><p><pre class="code-block">import { useMouse, usePreferredDark } from &apos;@vueuse/core&apos;
// tracks mouse position
const { x, y } = useMouse()
// is the user prefers dark theme
const isDark = usePreferredDark()</pre></p>"
`;

exports[`content-rich > code frame 2 1`] = `
"<p dir="auto">
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/webtoo.ls/@antfu"
      ><bdi>@<span>antfu</span></bdi></a
    ></span
  >
  Testing<br />
  <pre class="code-block">const a = hello</pre>
</p>
"
`;

exports[`content-rich > code frame empty 1`] = `"<p dir="auto"><pre class="code-block"></pre><br></p>"`;

exports[`content-rich > code frame no lang 1`] = `"<p dir="auto"><pre class="code-block">hello world</pre><br>no lang</p>"`;

exports[`content-rich > collapse mentions 1`] = `
"<p dir="auto">
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@elk"
      ><bdi>@<span>elk</span></bdi></a
    ></span
  >
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@elk"
      ><bdi>@<span>elk</span></bdi></a
    ></span
  >
  content
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@antfu"
      ><bdi>@<span>antfu</span></bdi></a
    ></span
  >
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/mastodon.roe.dev/@daniel"
      ><bdi>@<span>daniel</span></bdi></a
    ></span
  >
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@sxzz"
      ><bdi>@<span>sxzz</span></bdi></a
    ></span
  >
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@patak"
      ><bdi>@<span>patak</span></bdi></a
    ></span
  >
  content
</p>
"
`;

exports[`content-rich > custom emoji 1`] = `
"Daniel Roe
<picture class="custom-emoji" alt=":nuxt:" data-emoji-id="nuxt" title="nuxt"
  ><source
    srcset="
      https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png
    "
    media="(prefers-reduced-motion: reduce)" />
  <img
    src="https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png"
    alt=":nuxt:"
    title="nuxt"
/></picture>
"
`;

exports[`content-rich > empty 1`] = `""`;

exports[`content-rich > group mention > html 1`] = `
"<p>
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@pilipinas@lemmy.ml"
      ><bdi
        ><span data-type="mention" data-id="pilipinas@lemmy.ml"
          >@pilipinas</span
        ></bdi
      ></a
    ></span
  >
</p>
"
`;

exports[`content-rich > handles formatting from servers 1`] = `
"<h1>Fedi HTML Support Survey</h1>
<p>Does the following formatting come through accurately for you?</p>
<p></p>
<ul>
  <li>This is an indented bulleted list (not just asterisks).</li>
  <li><strong>This line is bold.</strong></li>
  <li><em>This line is italic.</em></li>
</ul>
<ol>
  <li>This list...</li>
  <li>...is numbered and indented</li>
</ol>
<h1>This line is larger.</h1>
"
`;

exports[`content-rich > handles html within code blocks 1`] = `
"<p dir="auto">
  HTML block code:<br />
  <pre class="code-block">
&lt;span class=&quot;icon--noto icon--noto--1st-place-medal&quot;&gt;&lt;/span&gt;
&lt;span class=&quot;icon--noto icon--noto--2nd-place-medal-medal&quot;&gt;&lt;/span&gt;</pre
  >
</p>
"
`;

exports[`content-rich > hashtag adds bdi 1`] = `
"<p dir="auto">
  Testing bdi is added
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/turkey"
        ><bdi>#<span>turkey</span></bdi></a
      ></VMenu
    ></span
  >
</p>
<p></p>
"
`;

exports[`content-rich > hashtag doesn't add 2 bdi 1`] = `
"<p dir="auto">
  Testing bdi not added
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/turkey"
        ><bdi></bdi></a></VMenu
  ></span>
</p>
<p></p>
"
`;

exports[`content-rich > hides collapsed mentions 1`] = `
"<p dir="auto">content</p>
"
`;

exports[`content-rich > inline code with link 1`] = `
"<p dir="auto">
  Inline code with link:
  <code
    >https://api.iconify.design/noto.css?icons=1st-place-medal,2nd-place-medal</code
  >
</p>
"
`;

exports[`content-rich > link + mention 1`] = `
"<p dir="auto">
  Happy
  <img
    src="/emojis/twemoji/1f917.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🤗"
  />
  we’re now using
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/webtoo.ls/@vitest"
      ><bdi>@<span>vitest</span></bdi></a
    ></span
  >
  (migrated from chai+mocha)
  <a
    href="https://github.com/ayoayco/astro-reactive-library/pull/203"
    rel="nofollow noopener noreferrer"
    target="_blank"
    ><span class="invisible">https://</span
    ><span class="ellipsis">github.com/ayoayco/astro-react</span
    ><span class="invisible">ive-library/pull/203</span></a
  >
</p>
"
`;

exports[`content-rich > p moved to div and text children replaced with p[dir="auto"] tags: br children removed 1`] = `
"<p dir="auto">
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/strangeobject.space/@bebatjof"
      ><bdi>@<span>bebatjof</span></bdi></a
    ></span
  >
  هذا اختبار:<br />أنا أحب الطريقة التي يتم بها دعم النموذج المزدوج العربي. تمت
  ترجمة الكلمة الأخيرة بشكل خاطئ وأحاول العثور على كيفية إصلاحها. أيضًا، يجب
  إصلاح نموذج 0.
</p>
<p></p>
<p dir="auto">
  This is a test:<br />I like how the arabic dual form is supported. The last
  one is mistranslated and I&#39;m trying to find how to fix it. Also, the form
  for 0 needs to be fixed.
</p>
<p></p>
"
`;

exports[`content-rich > plain text 1`] = `
"hello there
"
`;

exports[`content-rich > root p includes dir="auto" attr when mixed content 1`] = `
"<p dir="auto">
  هذا اختبار جديد
  <img
    src="/emojis/twemoji/1f426.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐦"
  /><img
    src="/emojis/twemoji/1f424.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐤"
  />
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/strangeobject.space/@bebatjof"
      ><bdi>@<span>bebatjof</span></bdi></a
    ></span
  >
  <br />أنا أحب الطريقة التي يتم بها دعم النموذج المزدوج العربي. تمت ترجمة
  الكلمة الأخيرة بشكل خاطئ وأحاول العثور على كيفية إصلاحها.
  <img
    src="/emojis/twemoji/1f426.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐦"
  /><img
    src="/emojis/twemoji/1f424.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐤"
  />
  ;). كما أن النموذج الخاص بـ 0 يحتاج إلى إصلاح
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/turkey"
        ><bdi>#<span>turkey</span></bdi></a
      ></VMenu
    ></span
  >
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/%D8%A7%D9%84%D8%B9%D8%B1%D8%A8%D9%8A%D8%A9"
        ><bdi>#<span>العربية</span></bdi></a
      ></VMenu
    ></span
  >
  .
</p>
<p></p>
<p dir="auto">
  This is a new test
  <img
    src="/emojis/twemoji/1f426.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐦"
  /><img
    src="/emojis/twemoji/1f424.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐤"
  />
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/strangeobject.space/@bebatjof"
      ><bdi>@<span>bebatjof</span></bdi></a
    ></span
  >
  <br />I like how the arabic dual form is supported. The last one is
  mistranslated and I&#39;m trying to find how to fix it.
  <img
    src="/emojis/twemoji/1f426.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐦"
  /><img
    src="/emojis/twemoji/1f424.svg"
    class="iconify-emoji iconify-emoji--twemoji"
    alt="🐤"
  />
  ;). Also, the form for 0 needs to be fixed
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/turkey"
        ><bdi>#<span>turkey</span></bdi></a
      ></VMenu
    ></span
  >
  <span
    ><VMenu
      placement="bottom-start"
      class="inline-block"
      close-on-content-click="false"
      ><a
        class="mention hashtag"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/tags/%D8%A7%D9%84%D8%B9%D8%B1%D8%A8%D9%8A%D8%A9"
        ><bdi>#<span>العربية</span></bdi></a
      ></VMenu
    ></span
  >
  .
</p>
<p></p>
"
`;

exports[`content-rich > shows some collapsed mentions grouped 1`] = `
"<p dir="auto">
  <mention-group
    ><span class="h-card"
      ><a
        class="u-url mention"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/@antfu"
        ><bdi>@<span>antfu</span></bdi></a
      ></span
    >
    <span class="h-card"
      ><a
        class="u-url mention"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/@patak"
        ><bdi>@<span>patak</span></bdi></a
      ></span
    >
    <span class="h-card"
      ><a
        class="u-url mention"
        rel="nofollow noopener noreferrer"
        to="/m.webtoo.ls/@sxzz"
        ><bdi>@<span>sxzz</span></bdi></a
      ></span
    ></mention-group
  >content
</p>
"
`;

exports[`content-rich > shows some collapsed mentions inline 1`] = `
"<p dir="auto">
  <span class="h-card"
    ><a
      class="u-url mention"
      rel="nofollow noopener noreferrer"
      to="/m.webtoo.ls/@antfu"
      ><bdi>@<span>antfu</span></bdi></a
    ></span
  >
  content
</p>
"
`;

exports[`editor > transform mentions 1`] = `
"
@elk Hello"
`;
